server/src/main/java/org/opensearch/tasks/TaskInfo.java

```diff
@@ -119,6 +129,12 @@ public TaskInfo(StreamInput in) throws IOException {
         startTime = in.readLong();
         runningTimeNanos = in.readLong();
         cancellable = in.readBoolean();
+        if (in.getVersion().onOrAfter(Version.V_2_0_0)) {
+            cancelled = in.readBoolean();
+        } else {
+            cancelled = false;
+        }
+        assert cancellable || cancelled == false : "uncancellable task cannot be cancelled";
```

Comment:
This might trigger in those cases where we are trying to cancel a task which cannot be cancelled when cancelled is true and cancellable is false. But as you mentioned we can throw a IllegalArgumentException instead of the assert error.